cache the generated test main file: dub_test_root.d

foerdi opened the pull request:
these changes prevent dub from rebuilding a project every time it runs a unittest.

wilzbach commented on changelog/cache-generated-test-config.dd:
```diff
@@ -0,0 +1,3 @@
+cache the generated test main file: dub_test_root.d
```
As this is part of the official changelog, It should be a bit more formal and abstract.
Have a look at the currently generated preview changelog:

http://dtest.dlang.io/artifact/website-f7458db425b639e3c2de12c651581f2157d84535-a2c0961de7a30ba66a2edcbbf5b69d6a/web/changelog/pending.html#cache-generated-test-config

foerdi replied:
Thank you for the review.

I try a different phrasing, unfortunate I am not a native english speaker, maybe you have a better suggestion?

wilzbach commented on source/dub/dub.d:
```diff
@@ -676,14 +693,22 @@ class Dub {
 					}
 					import_modules ~= dub.internal.utils.determineModuleName(lbuildsettings, NativePath(file), m_project.rootPackage.path);
+
+					auto ftime = getFileInfo(file).timeModified;
```
No need to query for file timestamps when regenerateMainFile is already true.

wilzbach commented on source/dub/dub.d:
```diff
@@ -662,6 +664,21 @@ class Dub {
+					settings.platform.platform.join("."),
+					settings.platform.architecture.join("."),
+					settings.platform.compiler, settings.platform.frontendVersion, lbuildsettings.sourceFiles.length);
+			}
```
1) I'm aware the code is old, but I would still move this into a separate method. This way it will be easier to add unittests for it.
2) Ideally you also move the bits to generate the package name from settings into a function instead of copying them around. This way it will be a lot easier in the future to find and adjust them
3) I'm aware the code is old, but new additions should try to be better and have a few basic comments for future readers.

foerdi replied:
```d
				mainfile = m_project.rootPackage.path ~ format(".dub/code/%s-%s-%s-%s-%s_%s_%s_dub_test_root.d", test_config, settings.buildType,
					settings.platform.platform.join("."),
					settings.platform.architecture.join("."),
					settings.platform.compiler, settings.platform.frontendVersion, lbuildsettings.sourceFiles.length);
			}
```

Be aware this is not the `.dub/build` folder. In this place I cannot access the `build_id` witch is generated in: https://github.com/dlang/dub/blob/3288bbb65282e1f47f106b195341ef4288c400ed/source/dub/generators/build.d#L330

The `buildsettings` can be changed in a project with dependencies.

Maybe someone has a better idea? I am aware this `.dub/code` solution is not ideal.

wilzbach replied:
> In this place I cannot access the build_id witch is generated in: 

What's stopping you from moving this out into a new public function:

```d
string computeBuildName(string config, GeneratorSettings settings)
{
    return format("%s-%s-%s-%s-%s_%s-%s", config, settings.buildType,
			settings.platform.platform.join("."),
			settings.platform.architecture.join("."),
			settings.platform.compiler, settings.platform.frontendVersion, hashstr);
}
```

And calling this from both places?

Furthermore, you would want to hash `lbuildsettings.sourceFiles` - not their length.

foerdi replied:
> Furthermore, you would want to hash lbuildsettings.sourceFiles - not their length.

This is a good idea, thanks.

wilzbach commented on source/dub/dub.d:
```diff
@@ -662,6 +664,21 @@ class Dub {
+	
+			mkdirRecurse(mainfile.parentPath.toNativeString());
+
+			auto mainfiletime = existsFile(mainfile) ? getFileInfo(mainfile).timeModified : SysTime.init;
```
This should not be necessary if `settings.force` is used.

wilzbach commented on source/dub/dub.d:
```diff
@@ -676,14 +693,22 @@ class Dub {
+
+					auto ftime = getFileInfo(file).timeModified;
+					if (ftime > Clock.currTime)
+						logWarn("File '%s' was modified in the future. Please re-save.", file);
```
```suggestion
						logWarn("File '%s' was modified in the future.", file);
```

I'm not sure how re-saving will help.

wilzbach reviewed (commented):
- [ ] should come with a test (or at least unittests for newly added code)

wilzbach commented on test/cache-generated-test-config/test.log:
```diff
@@ -0,0 +1 @@
+[ERROR] ./cache-generated-test-config.sh:13 DUB generated more then one main file
```
Accidental add?

wilzbach commented on test/cache-generated-test-config.sh:
```diff
@@ -0,0 +1,15 @@
+MAIN_FILES_COUNT=$(ls .dub/code/*dub_test_root.d | wc -l)
+
+[ "$MAIN_FILES_COUNT" != "1" ] && die $LINENO 'DUB generated more then one main file'
+[ "$EXECUTABLE_TIME" != $(stat -c "%Y" cache-generated-test-config-test-library) ] && die $LINENO 'The executable has been changed'
```
```suggestion
[ "$EXECUTABLE_TIME" != $(stat -c "%Y" cache-generated-test-config-test-library) ] && die $LINENO 'The executable has been rebuilt'
```

wilzbach commented on test/cache-generated-test-config.sh:
```diff
@@ -0,0 +1,15 @@
+
+[ "$MAIN_FILES_COUNT" != "1" ] && die $LINENO 'DUB generated more then one main file'
+[ "$EXECUTABLE_TIME" != $(stat -c "%Y" cache-generated-test-config-test-library) ] && die $LINENO 'The executable has been changed'
+[ "$MAIN_TIME" != $(stat -c "%Y" "$(ls .dub/code/*dub_test_root.d)") ] && die $LINENO 'The main file has been changed'
```
```suggestion
[ "$MAIN_TIME" != $(stat -c "%Y" "$(ls .dub/code/*dub_test_root.d | head -n1)") ] && die $LINENO 'The test main file has been rebuilt'
```

wilzbach commented on changelog/cache-generated-test-config.dd:
```diff
@@ -0,0 +1,4 @@
+Caching the generated unittest main file.
```
```suggestion
Caching of generated unittest runner (dub test)
```

wilzbach commented on changelog/cache-generated-test-config.dd:
```diff
@@ -0,0 +1,4 @@
+Caching the generated unittest main file.
+
+Prevent dub from rebuilding a unittest on a secondary `dub test` with out file changes.
+For dub projects without a user defined unittest configuration.
```
```suggestion
For projects without a user-defined unittest configuration `dub test` generates a main file automatically.
This main file is now being cached and won't be regenerated won subsequent runs without file changes.
```

(though to be fair, this feature won't affect many people as typically you don't rerun the testsuite without making changes to a file)

foerdi replied:
Thank you

foerdi commented:
Should I click on `Resolve conversation` after I pushed changes to the topic or the reviewer resolve the conversation after he is satisfied?

wilzbach commented:
> Should I click on `Resolve conversation` after I pushed changes to the topic or the reviewer resolve the conversation after he is satisfied?

Yes, please feel free to if it was a simple fix. Maybe be a bit more careful with a more in-depth discussion that wasn't fully resolved or might be relevant to other contributors so that they don't point out the same thing.

foerdi commented:
Did you have a idea why the osx test runner are falling?

```
[ERROR] /Users/runner/work/dub/dub/test/cache-generated-test-config.sh:8 command failed
[ERROR] /Users/runner/work/dub/dub/test/cache-generated-test-config.sh:8 command failed
[ERROR] Script failure.
```
test/cache-generated-test-config.sh:8:
```
EXECUTABLE_TIME="$(stat cache-generated-test-config-test-library | grep Modify)"
```

wilzbach commented:
> Did you have a idea why the osx test runner are failing?
> 
> ```
> [ERROR] /Users/runner/work/dub/dub/test/cache-generated-test-config.sh:8 command failed
> [ERROR] /Users/runner/work/dub/dub/test/cache-generated-test-config.sh:8 command failed
> [ERROR] Script failure.
> ```
> test/cache-generated-test-config.sh:8:
> ```
> EXECUTABLE_TIME="$(stat cache-generated-test-config-test-library | grep Modify)"
> ```

No, sorry but I recommend printing the output of stat on failure, s.t. you know what goes on.

foerdi commented:
Now I have reduced and simplified my code. Theoretically, the time check of changed code files for the main file generator is not necessary due to the module name hashes. So I removed the time check.

I noticed that a single macOS test run failed, but different from the last time, and it looks like this is not related to my changes (?). Can i do something about it?

foerdi commented:
It would be nice if someone could take the time to review my MR.
Many Thanks

Geod24 commented:
This seems to have introduced a regression, as the Alpine Linux packages now fail to bootstrap:
<img width="1143" alt="Screen Shot 2021-01-04 at 9 58 43" src="https://user-images.githubusercontent.com/2180215/103493271-9f8dad80-4e73-11eb-93cf-6ebf3182c9e3.png">

foerdi commented:
> This seems to have introduced a regression, as the Alpine Linux packages now fail to bootstrap:

I need some more Information:

Which package is failing to build or test?
How I can reproduce this error?

Maybe, can you check which files are in `.dub/code/*`?

Geod24 commented:
I didn't reduce the error yet, as there are other failures. Note that we're building with `gdc`, so it might be a GDC-specific issue cropping up. The MR is [here](https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/16347) if you want to take a look. Otherwise I'll get back to it ASAP (so probably this week).

foerdi commented:
Sorry for my late reply, I forgot to reply about my hollidaies. If this is still a problem, it would be great if you could reduce this error a bit more then I could help you. I couldn't figure out which project couldn't be tested.

Geod24 commented:
@foerdi : It's the check step of `dub` itself. Just running `bin/dub test --compiler=ldc2 -v` after building `dub` will show this.

Geod24 commented:
@foerdi : I don't know exactly how to reproduce this, as it's only triggered during the package build process.
What I see is that 2 values are created, one contains the `$DFLAGS`, and the other nothing, so this also changes the hash.
I solved it be reverting this PR in Alpine Linux.

Geod24 commented:
@foerdi : I just reproduced this accidentally. Seems to trigger with `DFLAGS` builds.
E.g. take https://github.com/Geod24/bitblob/ which is a very simple library.
I tried to build it with:
```console
$ DFLAGS='-wi' dub test -v
Using dub registry url 'https://code.dlang.org/'
Refreshing local packages (refresh existing: true)...
Looking for local package map at /var/lib/dub/packages/local-packages.json
Looking for local package map at /Users/geod24/.dub/packages/local-packages.json
Try to load local package map at /Users/geod24/.dub/packages/local-packages.json
Note: Failed to determine version of package custom-tool at .. Assuming ~master.
Looking for local package map at /Users/geod24/projects/geod24/bitblob/.dub/packages/local-packages.json
Determined package version using GIT: bitblob 1.1.4+commit.5.g73eaab4
Refreshing local packages (refresh existing: false)...
Looking for local package map at /var/lib/dub/packages/local-packages.json
Looking for local package map at /Users/geod24/.dub/packages/local-packages.json
Try to load local package map at /Users/geod24/.dub/packages/local-packages.json
Looking for local package map at /Users/geod24/projects/geod24/bitblob/.dub/packages/local-packages.json
Refreshing local packages (refresh existing: false)...
Looking for local package map at /var/lib/dub/packages/local-packages.json
Looking for local package map at /Users/geod24/.dub/packages/local-packages.json
Try to load local package map at /Users/geod24/.dub/packages/local-packages.json
Looking for local package map at /Users/geod24/projects/geod24/bitblob/.dub/packages/local-packages.json
Generating test runner configuration 'bitblob-test-library' for 'library' (library).
Get module name from path: /Users/geod24/projects/geod24/bitblob/source/geod24/bitblob.d
Refreshing local packages (refresh existing: false)...
Looking for local package map at /var/lib/dub/packages/local-packages.json
Looking for local package map at /Users/geod24/.dub/packages/local-packages.json
Try to load local package map at /Users/geod24/.dub/packages/local-packages.json
Looking for local package map at /Users/geod24/projects/geod24/bitblob/.dub/packages/local-packages.json
Configuring dependent bitblob, deps:
Performing "$DFLAGS" build using dmd for x86_64.
Target '/Users/geod24/projects/geod24/bitblob/.dub/build/bitblob-test-library-$DFLAGS-posix.osx.darwin-x86_64-dmd_v2.095.0-18ADA2291157B839D639A99D4340E461/bitblob-test-library' doesn't exist, need rebuild.
bitblob 1.1.4+commit.5.g73eaab4: building configuration "bitblob-test-library"...
dmd -c -of.dub/build/bitblob-test-library-$DFLAGS-posix.osx.darwin-x86_64-dmd_v2.095.0-18ADA2291157B839D639A99D4340E461/bitblob-test-library.o -wi -w -version=Have_bitblob -Isource/ .dub/code/bitblob-test-library--posix.osx.darwin-x86_64-dmd_v2.095.0-BB10D44F4B259993F6B84BBEB68A1E0C_dub_test_root.d source/geod24/bitblob.d -vcolumns
Error: module bitblob-test-library--posix.osx.darwin-x86_64-dmd_v2.095.0-BB10D44F4B259993F6B84BBEB68A1E0C_dub_test_root is in file '.dub/code/bitblob-test-library--posix.osx.darwin-x86_64-dmd_v2.095.0-BB10D44F4B259993F6B84BBEB68A1E0C_dub_test_root.d' which cannot be read
import path[0] = source/
import path[1] = /usr/local/opt/dmd/include/dlang/dmd
FAIL .dub/build/bitblob-test-library-$DFLAGS-posix.osx.darwin-x86_64-dmd_v2.095.0-18ADA2291157B839D639A99D4340E461/ bitblob-test-library executable
dmd failed with exit code 1.
$ ll .dub/code
total 8
-rw-r--r--  1 geod24  staff  555 Feb 26 18:36 bitblob-test-library-$DFLAGS-posix.osx.darwin-x86_64-dmd_v2.095.0-BB10D44F4B259993F6B84BBEB68A1E0C_dub_test_root.d
```

foerdi commented:
@Geod24 : Thank you, for your work and the reduced Test Case. Hopefully, I find some time this weekend to fix this bug.
 I have already an idea why this could be happening.

Now, how is the correct workflow?
Should I file a bug and a linked pull request, is a simple pull-request enough or can I recycle this pull-request / branch?

foerdi commented:
@Geod24 Today I found some time and fixed your issue: https://github.com/dlang/dub/pull/2116. Thank you for the discovery.
